tensorflow/contrib/rnn/kernels/lstm_ops.cc

```diff
@@ -0,0 +1,599 @@
+          TensorShape({input_size + cell_size_, cell_size_ * 4}),
+          &w_grad_tensor));
+    TensorMemZero(w_grad_tensor, stream);
+
+    Tensor* b_grad_tensor = nullptr;
+    OP_REQUIRES_OK(ctx, ctx->allocate_output("b_grad",
+          TensorShape({cell_size_ * 4}), &b_grad_tensor));
+    TensorMemZero(b_grad_tensor, stream);
+
+    for (int64 t = 0; t < sequence_len_max_; ++t) {
+      Tensor* x_grad_tensor = nullptr;
+      x_grad_tensors.allocate(
+          t, TensorShape({batch_size, input_size}), &x_grad_tensor);
+      TensorMemZero(x_grad_tensor, stream);
+    }
+
+    Tensor xh_tensor;
+    OP_REQUIRES_OK(ctx, ctx->allocate_temp(
+          DT_FLOAT, TensorShape({batch_size, input_size + cell_size_}),
+          &xh_tensor));
+
+    Tensor xh_grad_tensor;
+    OP_REQUIRES_OK(ctx, ctx->allocate_temp(DT_FLOAT,
+        TensorShape({batch_size, input_size + cell_size_}), &xh_grad_tensor));
+
+    Tensor states_grad_tensor;
+    OP_REQUIRES_OK(ctx, ctx->allocate_temp(DT_FLOAT,
+        TensorShape({batch_size, state_size}), &states_grad_tensor));
+    TensorMemZero(&states_grad_tensor, stream);
+
+    Tensor states_prev_grad_tensor;
+    OP_REQUIRES_OK(ctx, ctx->allocate_temp(DT_FLOAT,
+        TensorShape({batch_size, state_size}), &states_prev_grad_tensor));
+
+    for (int64 t = sequence_len_max - 1; t >= 0; --t) {
+      const Tensor& x_tensor = x_tensors[t];
+      const Tensor& states_prev_tensor =
+          t <= 0 ? *initial_state_tensor : states_tensors[t - 1];
+      const Tensor& states_tensor = states_tensors[t];
+      const Tensor& h_grad_tensor = h_grad_tensors[t];
+
+      Tensor* x_grad_tensor = x_grad_tensors[t];
+      const Tensor& states_grad_const_tensor = states_grad_tensor;
+
+      functor::LSTMCellBlockBprop<Device, USE_CUBLAS>()(
+          stream, ctx->eigen_device<Device>(),
+          batch_size, input_size, cell_size_,
+          x_tensor.matrix<float>(),
+          xh_tensor.matrix<float>(),
+          states_prev_tensor.matrix<float>(),
+          w_tensor->matrix<float>(),
+          b_tensor->vec<float>(),
+          states_tensor.matrix<float>(),
+          h_grad_tensor.matrix<float>(),
+          states_grad_const_tensor.matrix<float>(),
+          xh_grad_tensor.matrix<float>(),
+          x_grad_tensor->matrix<float>(),
+          states_prev_grad_tensor.matrix<float>(),
+          w_grad_tensor->matrix<float>(),
+          b_grad_tensor->vec<float>());
+
+      if (stream) {
+#if GOOGLE_CUDA
+        auto states_grad_ptr =
+            AsDeviceMemory(states_grad_tensor.flat<float>().data());
+        auto states_prev_grad_ptr =
+            AsDeviceMemory(states_prev_grad_tensor.flat<float>().data());
+        CHECK(stream->ThenMemcpy(
+            &states_grad_ptr, states_prev_grad_ptr,
+            states_grad_tensor.TotalBytes()).ok());
+#endif  // GOOGLE_CUDA
+      } else {
+        std::memcpy(states_grad_tensor.flat<float>().data(),
```

Comment:
FYI this might be slow since it is not a multi-threaded copy.  maybe consider using Eigen to do the copy inside the functor?